include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,383 @@
+        const vector3_type localH = ndf.generateH(upperHemisphereV, u.xy);
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
+
+        const scalar_type VdotH = hlsl::dot(V.getDirection(), H);
+        const scalar_type reflectance = _f(hlsl::abs(VdotH))[0];
+
+        scalar_type rcpChoiceProb;
+        scalar_type z = u.z;
+        bool transmitted = math::partitionRandVariable(reflectance, z, rcpChoiceProb);
+
+        Refract<scalar_type> r = Refract<scalar_type>::create(V.getDirection(), H);
+        bxdf::ReflectRefract<scalar_type> rr;
+        rr.refract = r;
+        ray_dir_info_type L = V.reflectRefract(rr, transmitted, rcpEta.value[0]);
+
+        const vector3_type T = interaction.getT();
+        const vector3_type B = interaction.getB();
+        const vector3_type _N = interaction.getN();
+
+        // fail if samples have invalid paths
+        const vector3_type Ldir = L.getDirection();
+        const scalar_type LdotH = hlsl::dot(Ldir, H);
+        if ((ComputeMicrofacetNormal<scalar_type>::isTransmissionPath(VdotH, LdotH) != transmitted) || (LdotH * hlsl::dot(_N, Ldir) < scalar_type(0.0)))
+            L.makeInvalid(); // should check if sample direction is invalid
+        else
+            cache = anisocache_type::create(VdotH, Ldir, H, T, B, _N, transmitted);
+
+        return sample_type::create(L, T, B, _N);
+    }
+    template<typename C=bool_constant<!IsAnisotropic>, typename D=bool_constant<!IsBSDF> >
+    enable_if_t<C::value && !IsAnisotropic && D::value && !IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, const vector2_type u, NBL_REF_ARG(isocache_type) cache)
+    {
+        anisocache_type aniso_cache;
+        sample_type s = generate(anisotropic_interaction_type::create(interaction), u, aniso_cache);
+        cache = aniso_cache.iso_cache;
+        return s;
+    }
+    template<typename C=bool_constant<!IsAnisotropic>, typename D=bool_constant<IsBSDF> >
+    enable_if_t<C::value && !IsAnisotropic && D::value && IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(isocache_type) cache)
+    {
+        anisocache_type aniso_cache;
+        sample_type s = generate(anisotropic_interaction_type::create(interaction), u, aniso_cache);
+        cache = aniso_cache.iso_cache;
+        return s;
+    }
+
+    template<class Interaction, class MicrofacetCache>
+    scalar_type __pdf(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        using quant_query_type = typename ndf_type::quant_query_type;
+        using dg1_query_type = typename ndf_type::dg1_query_type;
+
+        dg1_query_type dq = ndf.template createDG1Query<Interaction, MicrofacetCache>(interaction, cache);
+
+        fresnel_type _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, interaction.getNdotV());
```

Comment:
you only need to reorient fresnels for BSDFs, see https://github.com/Devsh-Graphics-Programming/Nabla/pull/930/files#r2401699310